tasks/resolve_reference.rb

```diff
@@ -0,0 +1,200 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require_relative '../../ruby_task_helper/files/task_helper.rb'
+require_relative '../../ruby_plugin_helper/lib/plugin_helper.rb'
+require 'net/http'
+require 'openssl'
+require 'json'
+require 'jwt'
+
+class GCloudInventory < TaskHelper
+  include RubyPluginHelper
+
+  AUTH_SCOPE        = 'https://www.googleapis.com/auth/compute.readonly'
+  AUTH_SKEW         = 60
+  CREDENTIALS_ENV   = 'GOOGLE_APPLICATION_CREDENTIALS'
+  CREDENTIALS_KEYS  = %w[client_email private_key project_id token_uri].freeze
+  GRANT_TYPE        = 'urn:ietf:params:oauth:grant-type:jwt-bearer'
+  SIGNING_ALGORITHM = 'RS256'
+
+  def resolve_reference(opts)
+    # Validate the target mapping
+    template = opts.delete(:target_mapping) || {}
+    unless template.key?(:uri) || template.key?(:name)
+      msg = "You must provide a 'name' or 'uri' in 'target_mapping' for the Google Cloud inventory plugin"
+      raise TaskHelper::Error.new(msg, 'bolt.plugin/validation-error')
+    end
+
+    # Retrieve credentials and access token for the API requests
+    creds = credentials(opts)
+    token = token(creds)
+    url   = "https://compute.googleapis.com/compute/v1/projects/#{opts[:project]}/zones/#{opts[:zone]}/instances"
+
+    # Build a list of compute engine instances, making multiple API requests if needed
+    instances = get_all_instances(url, token)
+
+    # Pull the necessary target data from each instance
+    attributes = required_data(template)
+    target_data = instances.map do |instance|
+      attributes.each_with_object({}) do |attr, acc|
+        attr = attr.first
+        acc[attr] = instance.key?(attr) ? instance[attr] : nil
+      end
+    end
```

Comment:
What does this step do? Can you not just call `apply_mapping` on each element of `instances` without first doing this top-level filter?

Reply:
That was lifted directly from `puppetlabs-ruby_plugin_helper`, which has an example from `puppetlabs-aws_inventory`. Removed the extra filtering and everything works.

Fyi, something wonky happened with the git history when I pushed changes and it auto-closed the PR. I've got a new one up here: https://github.com/puppetlabs/puppetlabs-gcloud_inventory/pull/4

Reply:
Ah I see. The difference in that example is that it's calling _methods_ on the instances rather than just doing hash lookups. So it's turning all the required fields into keys in a map (which is the structure required by `apply_mapping`), but in this version they're already keys in a map so that step is unnecessary.